File: src/scripts/run_frontier.ts
// @ts-nocheck — Depends on unimplemented modules (graph6, cycle_finder)
/**
 * Sprint 21: General Solver Pipeline — Streaming Frontier Verifier
 *
 * Streams non-isomorphic graphs from nauty/geng directly into a TypeScript
 * CycleFinder via Bun's native subprocess API. This decouples graph generation
 * (C-level symmetry breaking) from constraint verification (bounded DFS).
 *
 * Usage:
 *   brew install nauty
 *   bun run src/scripts/run_frontier.ts 10
 *   for n in {10..18}; do bun run src/scripts/run_frontier.ts $n; done
 */

import { Graph6Decoder } from "../math/graph6";
import { CycleFinder } from "../math/cycle_finder";
import * as fs from "node:fs/promises";
import { join } from "node:path";

const targetN = parseInt(process.argv[2] || "10", 10);

async function verifyFrontier(n: number) {
  console.log("═══════════════════════════════════════════════");
  console.log("  🔬 ERDŐS-GYÁRFÁS GENERAL SOLVER PIPELINE");
  console.log(`  n = ${n} vertices, min degree ≥ 3`);
  console.log("  Engine: nauty/geng (C) → Graph6 → CycleFinder (TS)");
  console.log("═══════════════════════════════════════════════\n");

  const startTime = performance.now();

  // Spawn geng: -d3 (min degree 3), -q (quiet), n vertices
  const proc = Bun.spawn(["geng", "-d3", "-q", n.toString()], {
    stdout: "pipe",
    stderr: "pipe",
  });

  const reader = proc.stdout.getReader();
  const decoder = new TextDecoder();

  let buffer = "";
  let graphsChecked = 0;
  let counterexample: string | null = null;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() || "";

    for (const line of lines) {
      const g6 = line.trim();
      if (!g6) continue;

      const adj = Graph6Decoder.decode(g6);
      graphsChecked++;

      if (!CycleFinder.hasPowerOfTwoCycle(adj)) {
        counterexample = g6;
        proc.kill();
        break;
      }

      if (graphsChecked % 50_000 === 0) {
        const elapsed = ((performance.now() - startTime) / 1000).toFixed(1);
        const rate = Math.round(graphsChecked / ((performance.now() - startTime) / 1000));
        console.log(`  ... ${graphsChecked.toLocaleString().padStart(12)} graphs verified (${rate.toLocaleString()}/s, ${elapsed}s) ...`);
      }
    }

    if (counterexample) break;
  }

  await proc.exited;

  const elapsed = ((performance.now() - startTime) / 1000).toFixed(2);
  const rate = Math.round(graphsChecked / ((performance.now() - startTime) / 1000));

  console.log("\n═══════════════════════════════════════════════");
  if (counterexample) {
    const adj = Graph6Decoder.decode(counterexample);
    const degrees = adj.map((a) => a.length);
    console.log(`  🚨 COUNTEREXAMPLE FOUND AT n=${n}`);
    console.log(`  Graph6: ${counterexample}`);
    console.log(`  Degrees: [${degrees}]`);
    console.log(`  Edges: ${adj.reduce((s, a) => s + a.length, 0) / 2}`);
  } else {
    console.log(`  ✅ VERIFIED n=${n}: All ${graphsChecked.toLocaleString()} graphs satisfy the conjecture.`);
    console.log(`  ⏱️  ${elapsed}s (${rate.toLocaleString()} graphs/s)`);
  }
  console.log("═══════════════════════════════════════════════\n");

  // Append result to JSONL log
  const result = {
    n,
    status: counterexample ? "counterexample" : "verified",
    graphs_checked: graphsChecked,
    elapsed_seconds: parseFloat(elapsed),
    rate_per_second: rate,
    counterexample_g6: counterexample,
    timestamp: new Date().toISOString(),
  };

  const dataDir = join(import.meta.dir, "../../data");
  await fs.mkdir(dataDir, { recursive: true });
  await fs.appendFile(
    join(dataDir, "frontier_results.jsonl"),
    JSON.stringify(result) + "\n",
  );
}

verifyFrontier(targetN).catch((err) => {
  console.error("💥 General Solver failed:", err);
  process.exit(1);
});

File: legacy/typescript/scripts/run_frontier.py
"""
Erdős-Gyárfás Frontier Pipeline: Nauty + Bounded DFS

Decouples graph generation (nauty/geng) from constraint verification (Python DFS).
This is the General Solver architecture: let C-level symmetry breaking handle
the combinatorial explosion, then verify each concrete graph in microseconds.

Usage:
    brew install nauty
    pip install networkx
    python3 src/scripts/run_frontier.py 10     # verify n=10
    for n in {10..18}; do python3 src/scripts/run_frontier.py $n; done  # overnight
"""

import subprocess
import networkx as nx
import time
import sys
import json
from pathlib import Path


def has_power_of_two_cycle(G):
    """
    Checks if the fixed graph G has a cycle of length 4, 8, or 16.
    Uses a highly optimized bounded DFS for specific lengths.
    """
    target_lengths = {4, 8, 16}  # For n <= 18, cycles can only be up to 16
    max_len = max(target_lengths)

    def dfs(start_node, current_node, depth, visited):
        if depth in target_lengths:
            if start_node in G[current_node]:
                return True
        if depth >= max_len:
            return False

        visited.add(current_node)
        for neighbor in G[current_node]:
            if neighbor not in visited:
                if dfs(start_node, neighbor, depth + 1, visited):
                    return True
        visited.remove(current_node)
        return False

    nodes = list(G.nodes())
    for i, start_node in enumerate(nodes):
        visited = set(nodes[:i])
        if dfs(start_node, start_node, 1, visited):
            return True

    return False


def verify_erdos_gyarfas(n):
    print(f"═══════════════════════════════════════════════")
    print(f"  🔬 ERDŐS-GYÁRFÁS FRONTIER — nauty/geng + DFS")
    print(f"  n = {n} vertices, min degree ≥ 3")
    print(f"═══════════════════════════════════════════════")
    start_time = time.time()

    # Call geng: -d3 (min degree 3), -q (quiet)
    cmd = ["geng", "-d3", "-q", str(n)]
    try:
        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, text=True)
    except FileNotFoundError:
        print("❌ geng not found. Install nauty: brew install nauty")
        sys.exit(1)

    graphs_checked = 0
    counterexample_found = False
    counterexample_g6 = None

    for line in process.stdout:
        g6_string = line.strip()
        if not g6_string:
            continue

        G = nx.from_graph6_bytes(g6_string.encode("ascii"))
        graphs_checked += 1

        if not has_power_of_two_cycle(G):
            print(f"\n  🚨 COUNTEREXAMPLE FOUND AT n={n} 🚨")
            print(f"  Graph6: {g6_string}")
            print(f"  Edges: {list(G.edges())}")
            print(f"  Degrees: {dict(G.degree())}")
            counterexample_found = True
            counterexample_g6 = g6_string
            process.kill()
            break

        if graphs_checked % 10000 == 0:
            elapsed = time.time() - start_time
            rate = graphs_checked / elapsed if elapsed > 0 else 0
            print(f"  ... {graphs_checked:>10,} graphs verified ({rate:,.0f}/s) ...")

    process.wait()
    elapsed = time.time() - start_time

    if not counterexample_found:
        print(f"\n  ✅ VERIFIED n={n}: All {graphs_checked:,} non-isomorphic graphs satisfy the conjecture.")
        print(f"  ⏱️  Time: {elapsed:.2f}s ({graphs_checked / elapsed:,.0f} graphs/s)")
    
    print(f"═══════════════════════════════════════════════\n")

    # Write results to data/
    result = {
        "n": n,
        "status": "counterexample" if counterexample_found else "verified",
        "graphs_checked": graphs_checked,
        "elapsed_seconds": round(elapsed, 2),
        "counterexample_g6": counterexample_g6,
    }
    
    data_dir = Path(__file__).parent.parent.parent / "data"
    data_dir.mkdir(exist_ok=True)
    results_file = data_dir / "frontier_results.jsonl"
    with open(results_file, "a") as f:
        f.write(json.dumps(result) + "\n")
    
    return not counterexample_found


if __name__ == "__main__":
    target_n = int(sys.argv[1]) if len(sys.argv) > 1 else 10
    verify_erdos_gyarfas(target_n)
